src/util/bson/bson.hpp

```diff
@@ -32,41 +32,670 @@
+    
+    // These are shortcuts for Bson(StringData(c_str)), and are
+    // needed to avoid unwanted implicit conversion of char* to bool.
+    Bson(char* c_str) noexcept
+        : Bson(StringData(c_str))
+    {
+    }
+    Bson(const char* c_str) noexcept
+        : Bson(StringData(c_str))
+    {
+    }
+
+    ~Bson() noexcept
+    {
+        switch (m_type) {
+            case STRING:
+                delete string_val;
+                string_val = NULL;
+                break;
+            case DOCUMENT:
+                delete document_val;
+                document_val = NULL;
+                break;
+            case ARRAY:
+                delete array_val;
+                array_val = NULL;
+                break;
+            case BINARY:
+                delete binary_val;
+                binary_val = NULL;
+                break;
+            case REGEX:
+                delete regex_val;
+                regex_val = NULL;
+                break;
+            default:
+                break;
+        }
+    }
+
+    Bson(Bson&& v) {
+        std::cout<<"TRYING TO MOVE BSON"<<std::endl;
+        m_type = v.m_type;
+    }
+
+    Bson(const Bson& v) {
+        m_type = v.m_type;
+        switch (v.m_type) {
+            case NONE:
+                break;
+            case INT32:
+                int32_val = v.int32_val;
+                break;
+            case INT64:
+                int64_val = v.int64_val;
+                break;
+            case BOOL:
+                bool_val = v.bool_val;
+                break;
+            case DOUBLE:
+                double_val = v.double_val;
+                break;
+            case TIME:
+                time_val = v.time_val;
+                break;
+            case DATE:
+                date_val = v.date_val;
+                break;
+            case OID:
+                oid_val = v.oid_val;
+                break;
+            case DECIMAL:
+                decimal_val = v.decimal_val;
+                break;
+            case MAX_KEY:
+                max_key_val = v.max_key_val;
+                break;
+            case MIN_KEY:
+                min_key_val = v.min_key_val;
+                break;
+            case BINARY:
+                binary_val = new std::vector<char>;
+                *binary_val = *v.binary_val;
+               break;
+            case REGEX:
+                regex_val = new RegularExpression;
+                *regex_val = *v.regex_val;
+                break;
+            case STRING:
+                string_val = new std::string;
+                *string_val = *v.string_val;
+                break;
+            case DOCUMENT:
+                document_val = new IndexedMap<Bson>;
+                *document_val = *v.document_val;
+                break;
+            case ARRAY:
+                array_val = new std::vector<Bson>;
+                *array_val = *v.array_val;
+                break;
+        }
+    }
+
+    Bson& operator=(Bson&& v) {
```

Comment:
The move assignment operator needs to be marked as `noexcept` or it won't actually be used in a lot of places, as the exception safety guarantees provided by things like `std::vector` aren't possible with a move assignment operator that throws exceptions.